Makefile

```diff
@@ -1,172 +1,383 @@
+# Extract GO version from go.mod file
+GO_VERSION                   ?= $(shell grep -E '^go' go.mod | awk {'print $$2'})
+# LDFLAGS
+GO_LDFLAGS                   += -X "github.com/mattermost/${APP_NAME}/internal/version.gitVersion=$(GIT_VERSION)"
+GO_LDFLAGS                   += -X "github.com/mattermost/${APP_NAME}/internal/version.gitCommit=$(APP_COMMIT)"
+GO_LDFLAGS                   += -X "github.com/mattermost/${APP_NAME}/internal/version.gitTreeState=$(GIT_TREESTATE)"
+GO_LDFLAGS                   += -X "github.com/mattermost/${APP_NAME}/internal/version.buildDate=$(BUILD_DATE)"
+
+
+# Architectures to build for
+GO_BUILD_PLATFORMS           ?= linux-amd64 linux-arm64 darwin-amd64 darwin-arm64 freebsd-amd64
+GO_BUILD_PLATFORMS_ARTIFACTS = $(foreach cmd,$(addprefix go-build/,${APP_NAME}),$(addprefix $(cmd)-,$(GO_BUILD_PLATFORMS)))
+
+# Build options
+GO_BUILD_OPTS                += -trimpath
+GO_TEST_OPTS                 += -v -timeout=180s
+# Temporary folder to output compiled binaries artifacts
+GO_OUT_BIN_DIR               := ./dist
+
+## Github Variables
+# A github access token that provides access to upload artifacts under releases
+GITHUB_TOKEN                 ?= a_token
+# Github organization
+GITHUB_ORG                   := mattermost
+# Most probably the name of the repo
+GITHUB_REPO                  := ${APP_NAME}
+
+# ====================================================================================
+# Colors
+
+BLUE   := $(shell printf "\033[34m")
+YELLOW := $(shell printf "\033[33m")
+RED    := $(shell printf "\033[31m")
+GREEN  := $(shell printf "\033[32m")
+CYAN   := $(shell printf "\033[36m")
+CNone  := $(shell printf "\033[0m")
+
+# ====================================================================================
+# Logger
+
+TIME_LONG	= `date +%Y-%m-%d' '%H:%M:%S`
+TIME_SHORT	= `date +%H:%M:%S`
+TIME		= $(TIME_SHORT)
+
+INFO = echo ${TIME} ${BLUE}[ .. ]${CNone}
+WARN = echo ${TIME} ${YELLOW}[WARN]${CNone}
+ERR  = echo ${TIME} ${RED}[FAIL]${CNone}
+OK   = echo ${TIME} ${GREEN}[ OK ]${CNone}
+FAIL = (echo ${TIME} ${RED}[FAIL]${CNone} && false)
+
+# ====================================================================================
+# Verbosity control hack
+
+VERBOSE ?= 0
+AT_0 := @
+AT_1 :=
+AT = $(AT_$(VERBOSE))
+
+# ====================================================================================
+# Targets
+
+help: ## to get help
+	@echo "Usage:"
+	@grep -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) |\
+	awk 'BEGIN {FS = ":.*?## "}; {printf "make ${CYAN}%-30s${CNone} %s\n", $$1, $$2}'
+
+.PHONY: build
+build: go-build-docker ## to build
+
+.PHONY: release
+release: build github-release ## to build and release artifacts
+
+.PHONY: package
+package: go-build package-software ## to build, package
+
+.PHONY: sign
+sign: docker-sign docker-verify ## to sign the artifact and perform verification
+
+.PHONY: lint
+lint: go-lint docker-lint ## to lint
+
+.PHONY: test
+test: go-test ## to test
```

Comment:
Is there a need to have `test` and `go-test` separately? Not a big deal, but would prefer to directly inline it here.

Reply:
This is intended to separate the testing logic from particular Makefile targets. We have a universal "test" target that can run various types of code tests for the application. For example if this application also needs some other testing, then a new target will be created "other-test", that will be part of the Test target. 

As described in the pull request's description, this concept is adapted from a Makefile template, and it's crafted to be versatile and expandable.

I hope this clarifies the purpose.

Reply:
Yep I understand. It felt like premature optimization to me. We are pre-creating a separate `go-test` target without actually knowing whether this project will ever have non-Go tests or not. But not a big deal either ways. Thanks Akis.

Reply:
The design was to offer the Makefile tamplete to developers, allowing them to perform any tests they needed. Our requirement is simply to have a "test" target that our CI system can invoke. Everything wull be decoupled under the "test" target. 

If you have strong objections to this approach, we can consider removing it, but I believe it's a beneficial design.